File: dcp-solutions/2024-10-03/solution.ex
```elixir
defmodule LockingBinaryTree do
  @enforce_keys [:value, :left, :right]

  defstruct value: nil, left: nil, right: nil, parent: nil

  def new(value) when is_integer(value), do: %LockingBinaryTree{value: value, left: nil, right: nil, parent: nil}
  
  @type t :: %__MODULE__{left: any(), right: any()}
  defstruct left: nil, right: nil
}

defmodule LockingBinaryTree.LockingNode do
  use LockingBinaryTree

  def is_locked(%LockingBinaryTree{}), do: false

  def lock(%LockingBinaryTree{}) when not is_locked(%LockingBinaryTree{}, []) do
    true
  else
    false
  end

  def unlock(%LockingBinaryTree{}) when not is_locked(%LockingBinaryTree{}, [node | _] = ancestors) do
    true
  else
    false
  end

  defp is_locked(node, ancestors), do: node[:value] in ancestors
end
```

File: dcp-solutions/2024-10-03/solution.c
```c
#include <stdbool.h>
typedef struct Node {
    int val;
    bool locked;
    struct Node* left;
    struct Node* right;
} Node;

struct Node* newNode(int key) {
    Node* node = (Node*) malloc(sizeof(Node));
    node->val = key;
    node->locked = false;
    node->left = NULL;
    node->right = NULL;
    return node;
}

bool isLocked(Node* node) {
    if (node == NULL) return false;

    if (node->locked) return true;

    Node* temp = node;
    while (temp != NULL) {
        if (temp->locked) return true;
        if (temp->left != NULL && temp->left->locked) return true;
        if (temp->right != NULL && temp->right->locked) return true;
        temp = temp->parent ? temp->parent : node;
    }

    return false;
}

bool lock(Node* node) {
    if (isLocked(node)) return false;

    if (node->locked) return true;

    if (node->left != NULL && isLocked(node->left)) return false;
    if (node->right != NULL && isLocked(node->right)) return false;

    node->locked = true;
    return true;
}

bool unlock(Node* node) {
    if (!node->locked) return false;

    if (isLocked(node->parent)) return false;

    node->locked = false;
    return true;
}
```